ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -22,28 +22,56 @@
 import java.util.ArrayList;
 import java.util.EnumSet;
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.NavigableMap;
-import java.util.Set;
+import java.util.Queue;
 import java.util.TreeMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
  * Exposes stats related to a {@link BlobStore} that is useful to different components.
  *
- * Note: This is the v0 implementation of BlobStoreStats. The v0 implementation walks through the entire index
- * and collect data needed to serve stats related requests.
+ * Note: This is the v1 implementation of BlobStoreStats. The v1 implementation periodically scans the index and create
+ * buckets to help serve requests within a forecast boundary. Stats related requests are either served via these buckets
+ * or a separate scan will be triggered to walk through the entire index and collect the requested data if the request
+ * is outside of the forecast boundary.
  */
 class BlobStoreStats implements StoreStats {
   static final String IO_SCHEDULER_JOB_TYPE = "BlobStoreStats";
   static final String IO_SCHEDULER_JOB_ID = "indexSegment_read";
+  private static final int ADD = 1;
+  private static final int SUBTRACT = -1;
+  private static final Logger logger = LoggerFactory.getLogger(BlobStoreStats.class);
 
   private final PersistentIndex index;
   private final Time time;
   private final DiskIOScheduler diskIOScheduler;
+  private final boolean isBucketingEnabled;
```

Comment:
StatsModule can still serve requests without enabling the bucketing optimization. `isStatsAvailable` sounds better since it indicates the state of the stats module without bucketing. However, might be weird/confusing to use `isStatsAvailable` within BlobStoreStats.